File: #2 LCS(continuous)/Codes/LCS.py
"""
CSU0018 Computer Algorithms
Programming Assignment 02
LCS (continuous) - Dynamic Programming
"""
def main(str_a, str_b) :
	last = ""
	sub_dict = {}

	for a_idx in range(len(str_a)):
		for b_idx in range(len(str_b)):
			if str_a[a_idx] == str_b[b_idx]:
				if not a_idx in sub_dict:
					sub_dict[a_idx] = {b_idx}
				else:
					sub_dict[a_idx].add(b_idx)
	
	for sub_len in range(2, min(len(str_a), len(str_b))):
		pop_list = []
		for (a_idx, idx_set) in sub_dict.items():
			if a_idx + sub_len > len(str_a):
				pop_list.append(a_idx)
				continue
			remove_list = []
			for b_idx in idx_set:
				if b_idx + sub_len > len(str_b):
					remove_list.append(b_idx)
					continue
				a_tmp = str_a[a_idx:a_idx + sub_len]
				b_tmp = str_b[b_idx:b_idx + sub_len]
				if a_tmp != b_tmp:
					remove_list.append(b_idx)
				else:
					last = a_tmp
			for rm_idx in remove_list:
				idx_set.remove(rm_idx)
		for key in pop_list:
			sub_dict.pop(key)		
	print(last)

#################################################################################
SAMPLE_NO = "1"
if __name__ == '__main__':
	main(*list(lines.strip('\n') for lines in open("test" + SAMPLE_NO + ".txt"))[1:])

File: #2 LCS(continuous)/Codes/LCS.cpp
/*
 * CSU0018 Computer Algorithms
 * Programming Assignment 02
 * LCS (continuous) - Dynamic Programming
 */
#include <iostream>
#include <string>
#include <cstdio>
#include <unordered_map>
#include <unordered_set>

inline size_t min(size_t a, size_t b) { return (a > b) ? b : a; }

void DP_print_LCS(std::string &str_a, std::string &str_b);

int main(int argc, char const *argv[]) {
	int n, m;
	while(scanf("%d %d", &n, &m) != EOF) {
		std::string str_a, str_b;
		std::cin >> str_a >> str_b;
		DP_print_LCS(str_a, str_b);
	}
	return 0;
}

void DP_print_LCS(std::string &str_a, std::string &str_b) {
	std::unordered_map<size_t, std::unordered_set<size_t>> sub_map;

	for(size_t a_idx = 0; a_idx < str_a.size(); ++a_idx)		// Fill map with sub string length = 1
		for(size_t b_idx = 0; b_idx < str_b.size(); ++b_idx)
			if(str_a[a_idx] == str_b[b_idx]) {
				if(sub_map.find(a_idx) == sub_map.end()) {
					std::unordered_set<size_t> idx_vec = { b_idx };
					sub_map.insert(std::make_pair(a_idx, idx_vec));
				}
				else
					sub_map[a_idx].emplace(b_idx);
			}

	std::string last = "";
	for(size_t sub_len = 1; sub_len < min(str_a.size(), str_b.size()); ++sub_len) {		// Eliminate
		std::unordered_set<size_t> pop_set;
		for(auto &pair : sub_map) {
			size_t a_idx = pair.first;
			if(a_idx + sub_len > str_a.size()) {
				pop_set.emplace(a_idx);
				continue;
			}

			std::unordered_set<size_t> remove_set;
			auto &idx_set = pair.second;
			for(auto b_idx : idx_set) {
				if(b_idx + sub_len > str_b.size()) {
					remove_set.emplace(b_idx);
					continue;
				}
				std::string a_tmp(str_a, a_idx, sub_len);
				std::string b_tmp(str_b, b_idx, sub_len);
				if(a_tmp != b_tmp)
					remove_set.emplace(b_idx);
				else
					last = a_tmp;
			}
			for(auto key : remove_set)
				idx_set.erase(key);
		}
		for(auto key : pop_set)
			sub_map.erase(key);
	}
	std::cout << last;
}